### PR TITLE
Update to GraalVM 24.0.0

### LINKED ISSUE #9609: Upgrade to GraalVM 24.0 to resolve `AssertionError`: no root node & co.
https://github.com/oracle/graal/issues/7131 is a fix made for us by the OracleLabs team. Without that fix we see errors like #9477. To benefit from the fix we shall upgrade to GraalVM 24.0:

> We tried to use the `@Bind("$node")` approach, but it seems to fail. Our [GraalVM version is 21.0.2](https://github.com/enso-org/enso/blob/380f5d2425a353ac1fa44351f31cf9140155fb29/build.sbt#L30) and our [Maven libraries are 23.1.2](https://github.com/enso-org/enso/blob/380f5d2425a353ac1fa44351f31cf9140155fb29/build.sbt#L34).

> This has only been fixed recently in 24.0. 

I guess we shall rather update to [24.0 packages](https://repo1.maven.org/maven2/org/graalvm/polyglot/polyglot/24.0.0/) released two weeks ago.

### PR DESCRIPTION
**Pull request by Akirathan**
Closes #9609

### Pull Request Description

Update the GraalVM-related Maven packages from **23.1.2** to **24.0.0**.

### Important Notes

- Reverted workarounds mentioned in https://github.com/enso-org/enso/issues/9477

The language home search functionality in Truffle changed. It used to find the home for Enso from the class loader of `EnsoLanguage`:
https://github.com/oracle/graal/blob/graal-23.1.2/truffle/src/com.oracle.truffle.polyglot/src/com/oracle/truffle/polyglot/LanguageCache.java#L362

In 24.0.0, the language home is no longer found via class loader, but via system properties:
https://github.com/oracle/graal/blob/graal-24.0.0/truffle/src/com.oracle.truffle.polyglot/src/com/oracle/truffle/polyglot/LanguageCache.java#L353

I provided a "fix" for this by an explicit specification of the `org.graalvm.language.enso.home` system property in our launchers.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

**Review by radeusgd — APPROVED** [review]

**Review by GregoryTravis — APPROVED** [review]

**Review by 4e6 — COMMENTED** [review]
Don't forget to update the Dockerfile

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArraySlice.java [hunk 1]
```diff
@@ -176,12 +176,12 @@ boolean isLimitReached(@Shared("warnsLib") @CachedLibrary(limit = "3") WarningsL
   }
 
   @ExportMessage
-  boolean hasType(@Shared("ignore") @Cached("1") int ignore) {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArraySlice.java:179]: Great. The upgrade pays off just by removing these `@Cached ignore` workarounds.

File: distribution/bin/enso [hunk 2]
```diff
@@ -8,5 +8,5 @@ for opt in "$@"; do
 done
 
 
-exec java --module-path $COMP_PATH $EXTRA_OPTS $JAVA_OPTS -m org.enso.runtime/org.enso.EngineRunnerBootLoader "$@"
+exec java --module-path $COMP_PATH -Dorg.graalvm.language.enso.home=$COMP_PATH $EXTRA_OPTS $JAVA_OPTS -m org.enso.runtime/org.enso.EngineRunnerBootLoader "$@"
```
Comment (JaroslavTulach) [inline distribution/bin/enso:11]: How exactly was this working before? It this option related to `RuntimeOptions.LANGUAGE_HOME_OVERRIDE`?

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArraySlice.java (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArraySlice.java:179]: Please remove also those 25 other usages of this workaround.

**Review by JaroslavTulach — APPROVED** [review]
Please remove other usages of `@Cached ignore`. Otherwise fine. Thank you for such a smooth upgrade.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArraySlice.java (hunk 1 above)
Comment (Akirathan) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArraySlice.java:179]: Removed the rest of the workarounds in 985a9cb6d0c860bca7e49ca058ef886a1b029c1e

Comment (Akirathan) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArraySlice.java:179]: Some leftovers addressed in 2744639c7271b7259c03025593ed139d21dcdb40 as well.

File: distribution/bin/enso (hunk 2 above)
Comment (Akirathan) [reply distribution/bin/enso:11]: > How exactly was this working before.

In most cases, Truffle used to find the home for use. See the PR description.

> Is this option related to `RuntimeOptions.LANGUAGE_HOME_OVERRIDE`?

Slightly. `org.graalvm.language.enso.home` system property is handled by Truffle, whereas `LANGUAGE_HOME_OVERRIDE` is handled by us and overrides any previous value. `LANGUAGE_HOME_OVERRIDE` is useful for testing, where we set it to point to the micro distribution.

**Comment by Akirathan** [comment]
> Don't forget to update the Dockerfile

@4e6 Unlike to previous GraalVM updates, updating the [Dockerfile](https://github.com/enso-org/enso/blob/b00dc9e9c00d09cc1f1c3d7749189e3f6864fea3/tools/ci/docker/Dockerfile) is not necessary anymore. That Dockerfile will need updating once we update JDK version, and not the maven packages. Note that this PR only updates associated Maven packages, not JDK itself.

**Review by 4e6 — APPROVED** [review]

**Pull request merged**